File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/settings/HandleCheckboxPreference.java
package com.lasthopesoftware.bluewater.settings;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.widget.CheckBox;
import com.lasthopesoftware.bluewater.client.stored.scheduling.SyncSchedulingWorker;

/**
 * Created by david on 11/14/15.
 */
class HandleCheckboxPreference {
	static void handle(Context context, String settingKey, CheckBox settingCheckbox) {
		settingCheckbox.setEnabled(false);
		final SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
		final boolean preference = sharedPreferences.getBoolean(settingKey, false);
		settingCheckbox.setChecked(preference);
		settingCheckbox.setOnCheckedChangeListener(
			(buttonView, isChecked) ->
			{
				sharedPreferences
					.edit()
					.putBoolean(settingKey, isChecked)
					.apply();

				SyncSchedulingWorker.scheduleSync(buttonView.getContext());
			});
		settingCheckbox.setEnabled(true);
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/settings/HandleCheckboxPreference.kt
package com.lasthopesoftware.bluewater.settings

import android.widget.CheckBox
import android.widget.CompoundButton
import com.lasthopesoftware.bluewater.settings.repository.ApplicationSettings
import com.lasthopesoftware.bluewater.settings.repository.access.HoldApplicationSettings

internal class HandleCheckboxPreference private constructor(private val applicationSettings: HoldApplicationSettings, private val updateSetting: (ApplicationSettings) -> (Boolean) -> Unit) : CompoundButton.OnCheckedChangeListener {
	override fun onCheckedChanged(buttonView: CompoundButton, isChecked: Boolean) {
		applicationSettings.promiseApplicationSettings()
			.then { s ->
				updateSetting(s)(isChecked)
				applicationSettings.promiseUpdatedSettings(s)
			}
	}

	companion object {
		fun handle(applicationSettings: HoldApplicationSettings, getSetting: (ApplicationSettings) -> Boolean, updateSetting: (ApplicationSettings) -> (Boolean) -> Unit, settingCheckbox: CheckBox) {
			settingCheckbox.isEnabled = false
			applicationSettings.promiseApplicationSettings()
				.then { s ->
					val preference = getSetting(s)
					settingCheckbox.isChecked = preference
					settingCheckbox.setOnCheckedChangeListener(
						HandleCheckboxPreference(applicationSettings, updateSetting))
					settingCheckbox.isEnabled = true
				}
		}
	}
}
